conda_smithy/configure_feedstock.py

```diff
@@ -459,6 +459,32 @@ def _collapse_subpackage_variants(
         if not meta.noarch:
             is_noarch = False
 
+    for var_dict in all_variants:
+        # on osx, merge MACOSX_DEPLOYMENT_TARGET & c_stdlib_version to max of either; see #1884
+        if var_dict.get("target_platform", "dummy").startswith("osx"):
+            # in global pinning, but use fallback to avoid having to set it in all tests
+            v_stdlib = var_dict.get("c_stdlib_version", "0.0")
+            # also in global pinning, but prepare for eventual disappearance
+            macdt = var_dict.get("MACOSX_DEPLOYMENT_TARGET", None)
+            if macdt is not None and v_stdlib != macdt:
+                # determine maximum version and use it to populate both
+                cond = VersionOrder(v_stdlib) > VersionOrder(macdt)
+                new_val = v_stdlib if cond else macdt
+                var_dict["c_stdlib_version"] = new_val
+                logger.warn(
+                    "Conflicting specification for minimum macOS deployment target!\n"
+                    "If your conda_build_config.yaml sets `MACOSX_DEPLOYMENT_TARGET`, "
+                    "please change the name of that key to `c_stdlib_version`!\n"
+                    f"Using {new_val}=max(c_stdlib_version, MACOSX_DEPLOYMENT_TARGET)."
+                )
+
+            # in any case, we set MACOSX_DEPLOYMENT_TARGET to match c_stdlib_version
+            # (if it is set), for ease of use in conda-forge-ci-setup
+            if "c_stdlib_version" in var_dict:
+                var_dict["MACOSX_DEPLOYMENT_TARGET"] = var_dict[
+                    "c_stdlib_version"
+                ]
+
```

Comment:
Overall, I think this could be fine.
I haven't reviewed the added tests.. but it'd probably be worth to try this out on a couple of feedstocks to be sure.
The code could be slightly simplified like this:
```suggestion
   for var_dict in all_variants:
        # on osx, merge MACOSX_DEPLOYMENT_TARGET & c_stdlib_version to max of either; see #1884
        if not var_dict.get("target_platform", "dummy").startswith("osx"):
            continue
        if "c_stdlib_version" not in var_dict:
            continue
        v_stdlib = var_dict["c_stdlib_version"]
        macdt = var_dict.get("MACOSX_DEPLOYMENT_TARGET", v_stdlib)
        if v_stdlib != macdt:
            # determine maximum version and use it to populate both
            if VersionOrder(v_stdlib) < VersionOrder(macdt):
                v_stdlib = macdt
            logger.warning(
                "Conflicting specification for minimum macOS deployment target!\n"
                "If your conda_build_config.yaml sets `MACOSX_DEPLOYMENT_TARGET`, "
                "please change the name of that key to `c_stdlib_version`!\n"
                f"Using {v_stdlib}=max(c_stdlib_version, MACOSX_DEPLOYMENT_TARGET)."
            )

        # we set MACOSX_DEPLOYMENT_TARGET to match c_stdlib_version, 
        # for ease of use in conda-forge-ci-setup
        var_dict["c_stdlib_version"] = v_stdlib
        var_dict["MACOSX_DEPLOYMENT_TARGET"] = v_stdlib

```
(Not much of a fan of the `v_stdlib`/`macdt` naming instead of just `c_stdlib_version`/`macosx_deployment_target`... but then again, the linter is set for a 79 char line limit which doesn't help :no_mouth: .)

FYI, I think from the `conda-build` side, `c_stdlib_version` doesn't even have to be an exact version; meaning, one might be able to set it to `>=10.13,<10.14` -- not saying that this would make much sense, but just that `VersionOrder(variant["c_stdlib_version"])` could raise an error (since it might be `VersionSpec` instead of just `VersionOrder`).
Not sure if we want/should handle this here for better error logs..

Reply:
Right. I guess in principle we should set the deployment target to the lower bound.

Reply:
> but then again, the linter is set for a 79 char line limit which doesn't help 😶 

Yeah, I've shortened the naming a couple of times because black just creates (IMO) garbage formatting otherwise.

> FYI, I think from the `conda-build` side, `c_stdlib_version` doesn't even have to be an exact version; meaning, one might be able to set it to `>=10.13,<10.14`

That's fair, and I don't mind raising an error here. Then again, conda-smithy is "The tool for managing conda-forge feedstocks.", and we're never going to set a range for the stdlib in the global pinning. So I really wouldn't mind that this case is not supported (it would be confusing at best, so I don't want to guess to use the lower bound; it should just error and we document how to use it properly).